File: backend/app/services/deduplication_service.py
import re
from typing import List, Dict, Any
from app.services.location_service import calculate_haversine

def normalize_text(text: str) -> str:
    """Normalize organization names by removing legal suffix tokens and punctuation."""
    t = (text or "").lower()
    t = re.sub(r"[^\w\s]", " ", t)
    # Remove corporate/hospital noise words
    t = re.sub(r"\b(pvt|ltd|limited|private|inc|corp|corporation|enterprise|llc|co|hospital|hospitals)\b", " ", t)
    return " ".join(t.split())

def string_similarity(s1: str, s2: str) -> float:
    """Compute token-based Jaccard similarity coefficient."""
    tokens1 = set(s1.split())
    tokens2 = set(s2.split())
    if not tokens1 or not tokens2:
        return 0.0
    intersection = tokens1.intersection(tokens2)
    union = tokens1.union(tokens2)
    return len(intersection) / len(union)

def deduplicate_records(records: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Merge duplicate records across sources using spatial proximity (<150m-300m)
    and name/domain/phone similarity.
    """
    deduped: List[Dict[str, Any]] = []

    for item in records:
        norm_name = normalize_text(item.get("name", ""))
        lat = item.get("latitude", 0.0)
        lon = item.get("longitude", 0.0)
        phone = item.get("phone", "")
        website = item.get("website", "")

        matched_index = -1

        for i, existing in enumerate(deduped):
            ex_norm = normalize_text(existing.get("name", ""))
            dist_km = calculate_haversine(lat, lon, existing.get("latitude", 0.0), existing.get("longitude", 0.0))

            # Rule 1: Very close physical distance (< 150m) and high name similarity
            if dist_km < 0.15 and string_similarity(norm_name, ex_norm) > 0.4:
                matched_index = i
                break

            # Rule 2: Same valid phone number within 1.0 km
            if phone and phone != "Not available" and phone == existing.get("phone") and dist_km < 1.0:
                matched_index = i
                break

            # Rule 3: Identical normalized web domain and same city
            if website and website != "Not available" and existing.get("website"):
                dom1 = website.replace("https://", "").replace("http://", "").split("/")[0]
                dom2 = existing.get("website", "").replace("https://", "").replace("http://", "").split("/")[0]
                if dom1 and dom1 == dom2 and dist_km < 2.0:
                    matched_index = i
                    break

        if matched_index >= 0:
            # Merge sources and retain highest verification score
            target = deduped[matched_index]
            if item.get("verification_score", 0) > target.get("verification_score", 0):
                # Update with higher score
                target["verification_score"] = item["verification_score"]
                target["verification_status"] = item["verification_status"]
            # Combine sources
            existing_sources = {s["url"]: s for s in target.get("verification_sources", [])}
            for s in item.get("verification_sources", []):
                existing_sources[s["url"]] = s
            target["verification_sources"] = list(existing_sources.values())
        else:
            deduped.append(item)

    return deduped

File: server/services/deduplicationService.ts
import { OrganizationItem } from '../../src/types.js';

/**
 * Calculates Haversine distance in meters between two lat/lon points
 */
export function calculateDistanceMeters(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
): number {
  const R = 6371e3; // Earth radius in metres
  const φ1 = (lat1 * Math.PI) / 180;
  const φ2 = (lat2 * Math.PI) / 180;
  const Δφ = ((lat2 - lat1) * Math.PI) / 180;
  const Δλ = ((lon2 - lon1) * Math.PI) / 180;

  const a =
    Math.sin(Δφ / 2) * Math.sin(Δφ / 2) +
    Math.cos(φ1) * Math.cos(φ2) * Math.sin(Δλ / 2) * Math.sin(Δλ / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return R * c;
}

export function normalizeOrgName(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^\w\s]/g, ' ')
    .replace(/\b(pvt|private|ltd|limited|llp|inc|corporation|corp|co|hospitals?|technologies|tech|solutions|services)\b/g, '')
    .replace(/\s+/g, ' ')
    .trim();
}

/**
 * Levenshtein distance ratio (0.0 to 1.0)
 */
export function stringSimilarity(str1: string, str2: string): number {
  const s1 = str1.trim();
  const s2 = str2.trim();
  if (s1 === s2) return 1.0;
  if (s1.length === 0 || s2.length === 0) return 0.0;

  const matrix: number[][] = [];
  for (let i = 0; i <= s1.length; i++) {
    matrix[i] = [i];
  }
  for (let j = 0; j <= s2.length; j++) {
    matrix[0][j] = j;
  }

  for (let i = 1; i <= s1.length; i++) {
    for (let j = 1; j <= s2.length; j++) {
      if (s1.charAt(i - 1) === s2.charAt(j - 1)) {
        matrix[i][j] = matrix[i - 1][j - 1];
      } else {
        matrix[i][j] = Math.min(
          matrix[i - 1][j - 1] + 1, // substitution
          matrix[i][j - 1] + 1,     // insertion
          matrix[i - 1][j] + 1      // deletion
        );
      }
    }
  }

  const distance = matrix[s1.length][s2.length];
  const maxLen = Math.max(s1.length, s2.length);
  return 1 - distance / maxLen;
}

/**
 * Checks if two organizations are duplicates using geographic distance,
 * normalized name similarity, phone match, or website match.
 */
export function areDuplicates(orgA: OrganizationItem, orgB: OrganizationItem): boolean {
  if (orgA.id === orgB.id) return true;
  if (orgA.type !== orgB.type) return false;

  const distanceMeters = calculateDistanceMeters(
    orgA.latitude,
    orgA.longitude,
    orgB.latitude,
    orgB.longitude
  );

  const normA = normalizeOrgName(orgA.name);
  const normB = normalizeOrgName(orgB.name);
  const nameSim = stringSimilarity(normA, normB);

  // Phone match check
  const phoneMatch =
    orgA.normalized_phone !== 'Not available' &&
    orgB.normalized_phone !== 'Not available' &&
    orgA.normalized_phone.replace(/\D/g, '') === orgB.normalized_phone.replace(/\D/g, '');

  // Website domain match check
  let domainMatch = false;
  try {
    if (orgA.website && orgB.website && orgA.website.startsWith('http') && orgB.website.startsWith('http')) {
      const domA = new URL(orgA.website).hostname.replace('www.', '');
      const domB = new URL(orgB.website).hostname.replace('www.', '');
      if (domA === domB) domainMatch = true;
    }
  } catch {
    // Ignore URL parse errors
  }

  // 1. Same phone number and within 500m -> duplicate
  if (phoneMatch && distanceMeters < 500) {
    return true;
  }

  // 2. High name similarity (> 0.85) and within 150m -> duplicate
  if (nameSim >= 0.82 && distanceMeters < 150) {
    return true;
  }

  // 3. Exactly identical domain and within 300m -> duplicate
  if (domainMatch && distanceMeters < 300) {
    return true;
  }

  // 4. Exact normalized name match and within 250m -> duplicate
  if (normA.length > 3 && normA === normB && distanceMeters < 250) {
    return true;
  }

  return false;
}

/**
 * Deduplicates a list of organization items, merging source links,
 * keeping the highest verification score, and retaining best address/phone.
 */
export function deduplicateOrganizations(items: OrganizationItem[]): OrganizationItem[] {
  const result: OrganizationItem[] = [];

  for (const item of items) {
    const existingIndex = result.findIndex((existing) => areDuplicates(existing, item));

    if (existingIndex === -1) {
      result.push({ ...item });
    } else {
      const target = result[existingIndex];
      // Merge sources without duplicates
      const mergedSources = [...target.verification_sources];
      for (const s of item.verification_sources) {
        if (!mergedSources.some((ms) => ms.url === s.url)) {
          mergedSources.push(s);
        }
      }
      target.verification_sources = mergedSources;

      // Keep better phone if missing
      if (target.phone === 'Not available' && item.phone !== 'Not available') {
        target.phone = item.phone;
        target.normalized_phone = item.normalized_phone;
        target.phone_type = item.phone_type;
      }

      // Keep better website if missing
      if ((!target.website || target.website === 'Not available') && item.website && item.website !== 'Not available') {
        target.website = item.website;
      }

      // Keep longer/better address
      if ((!target.address || target.address === 'Not available') && item.address && item.address !== 'Not available') {
        target.address = item.address;
      }

      // Keep higher verification score
      if (item.verification_score > target.verification_score) {
        target.verification_score = item.verification_score;
        target.verification_status = item.verification_status;
        target.verification_breakdown = item.verification_breakdown;
      }
    }
  }

  return result;
}
